rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,513 @@
+    mapAttrs = f: s: #...
+```
+
+### `/** */` inspired by the current multiline strings 
+
+Example of an alternative multiline`/** */` format; with `Markdown` Headings.
+
+TODO: Work out the details for multiline doc comments together with documentation-team. This section might be removed and included into the "detailed design" section.
+
+```nix
+    /** 
+        mapAttrs is a function in lib
+     
+        # Example
+    
+        ```
+        comprehensive example
+        shows how to use this
+        ```
+     
+        # Type
+        
+        foo -> bar 
+    */
+    mapAttrs = f: s: #...
+```
+
+
+## Javadoc style
+
+Example of the discarded `javadoc` format.
+
+```java
+  /**
+    * A short description
+    * @author  Stefan Schneider
+    * @version 1.1
+    * @see    https://some.url
+    */
+    public class Product {
+     ...
+    }
+```
+
+## Alternative approach - just comments
+
+There is the idea from python that doc-comments are just strings, not even special ones. Strings will be docstrings if they follow specific placement rules. However, we thought this was a bad idea to follow. Such complex placement rules require the users to understand where those places are; with nix syntax, this is slightly more complex than with python. Because we don't have keywords such as `class MyClass():` or `def function():` where placement would be obvious
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- `nixodc` offers comments to describe function arguments. This is currently not compatible until some sections for `args` are defined.
+
+- Will `nix` itself implement native support like in rust -> `cargo doc`
+
+- How can a tool keep the connection from where a docstring was defined and where the attribute was exposed (lib/default. nix exposes mapAttrs which is defined at lib/attrsets.nix)
+  - There are more complicated things.
+
+-> Answer: A Tool might be able to keep track of a percentage of expressions, and sometimes it may be very hard or impossible. For that case, the doc-comment can offer a dedicated Keyword to override the scope.
+
+e.g.
+
+The following is an idea for a problem that will arise if tools try to track doc-comments' positions and the location in the nixpkgs tree. (Although this problem is not nixpkgs specific)
+
+```nix
+## This file is called somewhere that cannot be automatically tracked/is impossible to analyze statically.
+## The 'TreePath' override can be used by the docstring author to set fixed paths in the nixpkgs expression.
+## This tells the doc-tool that the expression is available in two aliases: `pkgs.stdenv` and `lib.sth` 
+## (This behavior will not be specified in this RFC)
+## # Meta
+## ## TreePath
+## - pkgs.stdenv
+## - lib.sth
+
+{...}:
+{
+    # returns something
+}
+```
+
+# Future work
+[future]: #future-work
+
+## Editor support
+
+When starting hitting {enter} inside a doc-block the new line, should be automatically prefixed with `##` accordingly.
```

Comment:
Is this somehow standardized?

Reply:
Couldnt find how rust lsp is doing it. But yes there should be some code action or something in the lsp world. Do you know how it works? It must be nested in some capability of the lsp

Reply:
Some research revealed:
- The lsp feature for the rust-language: https://github.com/rust-lang/vscode-rust/pull/156 
- Issue for vscode to implement that natively: https://github.com/microsoft/vscode/issues/26694

Reply:
In other words, LSP standard does not tell a thing about mandatory keystrokes for inserting programming constructs.
Phew!